File: tests/example-program-gen/exp/instructions/initializeWithValues2.ts
import { TransactionInstruction, PublicKey } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface InitializeWithValues2Args {
  vecOfOption: Array<BN | null>
}

export interface InitializeWithValues2Accounts {
  state: PublicKey
  payer: PublicKey
  systemProgram: PublicKey
}

export const layout = borsh.struct([
  borsh.vec(borsh.option(borsh.u64()), "vecOfOption"),
])

export function initializeWithValues2(
  args: InitializeWithValues2Args,
  accounts: InitializeWithValues2Accounts
) {
  const keys = [
    { pubkey: accounts.state, isSigner: true, isWritable: true },
    { pubkey: accounts.payer, isSigner: true, isWritable: true },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([248, 190, 21, 97, 239, 148, 39, 181])
  const buffer = Buffer.alloc(1000)
  const len = layout.encode(
    {
      vecOfOption: args.vecOfOption,
    },
    buffer
  )
  const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len)
  const ix = new TransactionInstruction({ keys, programId: PROGRAM_ID, data })
  return ix
}

File: tests/client_gen/example_program_gen/instructions/initialize_with_values2.py
from __future__ import annotations

import typing

import borsh_construct as borsh
from construct import Construct
from solana.publickey import PublicKey
from solana.system_program import SYS_PROGRAM_ID
from solana.transaction import AccountMeta, TransactionInstruction

from ..program_id import PROGRAM_ID


class InitializeWithValues2Args(typing.TypedDict):
    vec_of_option: list[typing.Optional[int]]


layout = borsh.CStruct(
    "vec_of_option" / borsh.Vec(typing.cast(Construct, borsh.Option(borsh.U64)))
)


class InitializeWithValues2Accounts(typing.TypedDict):
    state: PublicKey
    payer: PublicKey


def initialize_with_values2(
    args: InitializeWithValues2Args,
    accounts: InitializeWithValues2Accounts,
    program_id: PublicKey = PROGRAM_ID,
    remaining_accounts: typing.Optional[typing.List[AccountMeta]] = None,
) -> TransactionInstruction:
    keys: list[AccountMeta] = [
        AccountMeta(pubkey=accounts["state"], is_signer=True, is_writable=True),
        AccountMeta(pubkey=accounts["payer"], is_signer=True, is_writable=True),
        AccountMeta(pubkey=SYS_PROGRAM_ID, is_signer=False, is_writable=False),
    ]
    if remaining_accounts is not None:
        keys += remaining_accounts
    identifier = b"\xf8\xbe\x15a\xef\x94'\xb5"
    encoded_args = layout.build(
        {
            "vec_of_option": args["vec_of_option"],
        }
    )
    data = identifier + encoded_args
    return TransactionInstruction(keys, program_id, data)
